src/wakefs/fuse.cpp

```diff
@@ -127,6 +129,32 @@ static bool collect_result_metadata(const std::string daemon_output, const struc
   return !result_ss.fail();
 }
 
+static void exec_in_pidns(pidns_args *nsargs) {
+  size_t stack_size = 1024 * 1024;
+  uint8_t *child_stack = (uint8_t *)malloc(stack_size) + stack_size;
```

Comment:
malloc assures 8-byte aligned pointers but the x86_64 abi requires 16-byte aligned pointers. We should bump the pointer to alignment.

Reply:
Ah, sadly that's not possible because glibc's clone makes modifications to the stack in the current address space before jumping into the child (which makes sense, it has to setup the arguments and return address etc...) so you have to give it some extra space. So I recommend `alignas(16) uint8_t child_stack[4096]` as that will protect you from any issues but it will still be on the stack so the child processes will get a plenty big stack (much bigger than 4096 bytes).

Another thing I got wrong but I see you got right is that you have to give the *last* byte because the stack grows towards lower addresses. You have an off by 1 error in yours though since the pointer you pass to clone has to be a valid byte, not just past the end of it. So add in a `+ sizeof(child_stack) - 1` but then you'll need to re-align it which means you have to do some pointer arithmetic...or you can just increase the size of child_stack by one so that its aligned after the subtraction.

```
alignas(16) uint8_t child_stack[4097];
void* stk = reinterpret_cast<void*>(child_stack + sizeof(child_stack) - 1);
pid_t child_pid = clone(pidns_init, stk, CLONE_NEWPID | SIGCHLD, nsargs);
```

I think that's all correct now.

Reply:
I don't think the `- 1` is necessary. The official [man page](https://man7.org/linux/man-pages/man2/clone.2.html) has an example at the very bottom that points to just past the allocation. I think the reason it's fine is normally the stack pointer points to the _most recent_ value pushed to the stack, and the contract for a function is for it to change the stack pointer and to only use the address up to and including the new stack pointer value.

Stack-allocating the stack for the child process might be a bit dangerous, since if you overflow the child process's stack, it could clobber the parent process's stack.

I was a bit concerned about the `malloc` as well at first, but after reading through the rest of the PR more carefully, I see that the process that calls `exec_in_pidns()` will never return since it always exits.

Reply:
Nothing will clobber because the memory is copy on write in the child process like it is with fork(). You have to specify CLONE_VM in order to get it to *not* copy on write. But you seem to be right about not needing the -1.

Reply:
I meant in the glibc `clone()` function, which does a bit of work in the current address space before actually calling the `clone` syscall.

Reply:
Right, yeah, I think I agree that that should be enough.

Reply:
Is the consensus: stay with malloc() but use 1 page as size?

Reply:
My understanding of the consensus was this. Having the stack allocated array be a page in size ensures that glibc won't clobber anything you care about and since you're using part of the stack the child processes stack will be close to the same size as the parent process's stack and there's no allocation.

```
alignas(16) uint8_t child_stack[4096];
void* stk = reinterpret_cast<void*>(child_stack + sizeof(child_stack));
pid_t child_pid = clone(pidns_init, stk, CLONE_NEWPID | SIGCHLD, nsargs);
```